### PR TITLE
Allow systems to be made inaccessible, preventing them from being seen or traveled to

### PR DESCRIPTION
**Pull request by Amazinite**
**Feature:** This PR is a reopening of #5752, which implements the feature request mentioned in #5734.

## Feature Details
Adds a new tag that systems can use, `inaccessible`, which prevents systems from being discovered on the map under any circumstances. Different to `hidden`, which only prevents systems from being viewed when not linked to.

This is a reopening of #5752, made up to date to remove conflicts. I've also removed the data changes to make use of this new tag, keeping the feature's creation and use separate. We can put this tag to use once it's available.

## UI Screenshots
Marked Sol as inaccessible, while keeping all link definitions. Sol cannot be seen, has no links to or from it, and cannot be selected in any way.
image

Marked Waypoint as inaccessible. The wormhole link cannot be seen and the wormhole cannot be landed on.
image
image

## Usage Examples
```
system Postverta
	inaccessible
```

## Testing Done
See the UI screenshot for the testing done.
Attempted to select Sol or otherwise reach it by various means: click on the map, hit tab or J on the map, hit J with no travel plan and while on the main panel. Unable to.
Observed fleet movements around Sol. Nothing came from or entered Sol.
Observed no jobs with Sol as a destination. Also created a mission which has a destination on Sol. Honestly not sure why, but the destination changes to the source, and that was before I even changed the LocationFilter to skip over invisible systems.
Tried to land on wormhole to Waypoint. Unable to.

### Automated Tests Added
N/A

## Performance Impact
N/A

### CODE REVIEW

**Comment by Amazinite** [comment]
Still need to check wormhole interactions with invisible systems. What I'd like is that wormholes don't draw links to invisible systems, and also don't even let you enter if the destination is an invisible system (but at the same time don't suddenly convert to being a normal landable object).

**Comment by Hurleveur** [comment]
Do we ever need the inaccessible links? Maybe Links() should only return the valid (accessible links) and then almost nothing needs to change?

**Comment by Amazinite** [comment]
I did actually start thinking about that toward the end. We'd prune `links` into an `accessibleLinks` vector that we'd probably just return via `Links()`. Then UpdateSystem would need to clear and update `accessibleLinks` whenever it's called. Since UpdateSystems is already called when the game finishes loading and every time a system is modified by an event, lots of stuff would be handled. We'd still need to check the inaccessible variable in a few locations (e.g. searching system names), but it would be much less than with what is done right now.

File: source/MapDetailPanel.cpp [hunk 1]
```diff
@@ -227,7 +228,6 @@ bool MapDetailPanel::KeyDown(SDL_Keycode key, Uint16 mod, const Command &command
 		{
 			// Skip the currently selected link, if any, and non valid system links. Also skip links to
 			// systems the player has not seen, and skip hyperspace links if the
-			// player has not visited either end of them.
 			if(!it->IsValid() || it == original)
```
Comment (Amazinite) [inline source/MapDetailPanel.cpp:231]:
```suggestion
			// player has not visited either end of them.
			if(!it->IsValid() || it == original)
```

File: source/MapPanel.cpp [hunk 2]
```diff
@@ -894,16 +903,17 @@ void MapPanel::UpdateCache()
 	for(const auto &it : GameData::Systems())
 	{
 		const System *system = &it.second;
-		if(!system->IsValid() || !player.HasSeen(*system))
+		if(!system->IsValid() || system->Inaccessible() || !player.HasSeen(*system))
```
Comment (Amazinite) [inline source/MapPanel.cpp:906]: Redundant check with the face that an inaccessible system will return an empty Links set.
```suggestion
		if(!system->IsValid() || !player.HasSeen(*system))
```

File: source/MapPanel.cpp [hunk 3]
```diff
@@ -894,16 +903,17 @@ void MapPanel::UpdateCache()
...
 		for(const System *link : system->Links())
 			if(link < system || !player.HasSeen(*link))
 			{
-				// Only draw links between two systems if one of the two is
+				// Only draw links between two visible systems if one of the two is
 				// visited. Also, avoid drawing twice by only drawing in the
 				// direction of increasing pointer values.
-				if((!player.HasVisited(*system) && !player.HasVisited(*link)) || !link->IsValid())
+				if((!player.HasVisited(*system) && !player.HasVisited(*link))
+						|| !link->IsValid() || link->Inaccessible())
```
Comment (Amazinite) [inline source/MapPanel.cpp:916]:
```suggestion
				if((!player.HasVisited(*system) && !player.HasVisited(*link)) || !link->IsValid())
```

File: source/MapPanel.cpp [hunk 4]
```diff
@@ -894,16 +903,17 @@ void MapPanel::UpdateCache()
...
 		const System *system = &it.second;
-		if(!system->IsValid() || !player.HasSeen(*system))
+		if(!system->IsValid() || system->Inaccessible() || !player.HasSeen(*system))
 			continue;
 
 		for(const System *link : system->Links())
 			if(link < system || !player.HasSeen(*link))
 			{
-				// Only draw links between two systems if one of the two is
+				// Only draw links between two visible systems if one of the two is
```
Comment (Amazinite) [inline source/MapPanel.cpp:912]:
```suggestion
				// Only draw links between two systems if one of the two is
```

**Review by Amazinite — COMMENTED** [review]
A few more cases were we don't need to check anymore.

File: source/LocationFilter.cpp [hunk 5]
```diff
@@ -468,8 +469,8 @@ const Planet *LocationFilter::PickPlanet(const System *origin, bool hasClearance
 	for(const auto &it : GameData::Planets())
 	{
 		const Planet &planet = it.second;
-		// Skip entries with incomplete data.
-		if(!planet.IsValid())
+		// Skip planets with incomplete data or which are from iaccessible systems.
```
Comment (Hurleveur) [inline source/LocationFilter.cpp:472]:
```suggestion
		// Skip planets with incomplete data or which are from inaccessible systems.
```

File: source/System.h [hunk 6]
```diff
@@ -196,12 +199,18 @@ class System {
...
+	// All possible hyperspace links to other systems.
 	std::set<const System *> links;
+	// Only those hyperspace links to other systems that are accessible.
+	std::set<const System *> accessibleLinks;
+	// Other systems that can be accessed from this system via a jump drive at various jump ranges.
 	std::map<double, std::set<const System *>> neighbors;
 
-	// Defines whether this system can be seen when not linked.
+	// Defines whether this system can be seen when not linked. A hidden system will
+	// not appear when in view range, only when linked to a visited system.
```
Comment (Hurleveur) [inline source/System.h:210]:
```suggestion
	// not appear when in view range, except when linked to a visited system.
```
is this what you meant?

**Review by Hurleveur — APPROVED** [review]
I can't find anything wrong with the code, and it seems like you thought of wormholes this should be good. That said I think we will need to update the teleport PR to take this into account later @RisingLeaf to prevent tping into inaccessible systems.

Storing it.second in variables at multiple places makes the code much more readable; and since the scope of this PR is so small I think its fine to keep them here like you did.

I just found a few things to fix in the comments

**Comment by RisingLeaf** [comment]
> I can't find anything wrong with the code, and it seems like you thought of wormholes this should be good. That said I think we will need to update the teleport PR to take this into account later @RisingLeaf to prevent tping into inaccessible systems.

Uh... I dont want to work, just merge teleport first and have @Amazinite work out the problem on this one

**Review by RisingLeaf — APPROVED** [review]
nice and clean

File: source/System.h [hunk 7]
```diff
@@ -96,8 +96,11 @@ class System {
 	// If this system has its own jump range, then it will always return the
 	// systems within that jump range instead of the jump range given.
 	const std::set<const System *> &JumpNeighbors(double neighborDistance) const;
-	// Whether this system can be seen when not linked.
+	// Defines whether this system can be seen when not linked. A hidden system will
+	// not appear when in view range, only when linked to a visited system.
```
Comment (Amazinite) [inline source/System.h:100]:
```suggestion
	// not appear when in view range, except when linked to a visited system.
```

File: source/System.cpp [hunk 8]
```diff
@@ -571,14 +590,23 @@ const set<const System *> &System::JumpNeighbors(double neighborDistance) const
 
 
 
-// Whether this system can be seen when not linked.
+// Defines whether this system can be seen when not linked. A hidden system will
+// not appear when in view range, only when linked to a visited system.
```
Comment (Amazinite) [inline source/System.cpp:594]:
```suggestion
// not appear when in view range, except when linked to a visited system.
```

**Pull request merged**